translator/trace/oc_to_otlp.go

```diff
@@ -0,0 +1,420 @@
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return consumerdata.OTLPTraceData{}
+	}
+
+	resource := ocNodeResourceToOtlp(td.Node, td.Resource)
+
+	resourceSpans := &otlptrace.ResourceSpans{
+		Resource: resource,
+	}
+	resourceSpanList := []*otlptrace.ResourceSpans{resourceSpans}
+
+	if len(td.Spans) != 0 {
+		resourceSpans.Spans = make([]*otlptrace.Span, 0, len(td.Spans))
+
+		for _, ocSpan := range td.Spans {
+			if ocSpan == nil {
+				// Skip nil spans.
+				continue
+			}
+
+			otlpSpan := ocSpanToOtlp(ocSpan)
+
+			if ocSpan.Resource != nil {
+				// Add a separate ResourceSpans item just for this span since it
+				// has a different Resource.
+				separateRS := &otlptrace.ResourceSpans{
+					Resource: ocNodeResourceToOtlp(nil, ocSpan.Resource),
+					Spans:    []*otlptrace.Span{otlpSpan},
+				}
+				resourceSpanList = append(resourceSpanList, separateRS)
+			} else {
+				// Otherwise add the span to the first ResourceSpans item.
+				resourceSpans.Spans = append(resourceSpans.Spans, otlpSpan)
+			}
+		}
+	}
+
+	return consumerdata.NewOTLPTraceData(resourceSpanList)
+}
+
+func timestampToUnixnano(ts *timestamp.Timestamp) uint64 {
+	return uint64(internal.TimestampToTime(ts).UnixNano())
+}
+
+func ocSpanToOtlp(ocSpan *octrace.Span) *otlptrace.Span {
+	attrs, droppedAttrCount := ocAttrsToOtlp(ocSpan.Attributes)
+	events, droppedEventCount := ocEventsToOtlp(ocSpan.TimeEvents)
+	links, droppedLinkCount := ocLinksToOtlp(ocSpan.Links)
+
+	childSpanCount := int32(0)
+	if ocSpan.ChildSpanCount != nil {
+		childSpanCount = int32(ocSpan.ChildSpanCount.Value)
+	}
+
+	otlpSpan := &otlptrace.Span{
+		TraceId:                ocSpan.TraceId,
+		SpanId:                 ocSpan.SpanId,
+		Tracestate:             ocTraceStateToOtlp(ocSpan.Tracestate),
+		ParentSpanId:           ocSpan.ParentSpanId,
+		Name:                   truncableStringToStr(ocSpan.Name),
+		Kind:                   ocSpanKindToOtlp(ocSpan.Kind, ocSpan.Attributes),
+		StartTimeUnixnano:      timestampToUnixnano(ocSpan.StartTime),
+		EndTimeUnixnano:        timestampToUnixnano(ocSpan.EndTime),
+		Attributes:             attrs,
+		DroppedAttributesCount: droppedAttrCount,
+		Events:                 events,
+		DroppedEventsCount:     droppedEventCount,
+		Links:                  links,
+		DroppedLinksCount:      droppedLinkCount,
+		Status:                 nil,
+		LocalChildSpanCount:    childSpanCount,
+	}
+
+	return otlpSpan
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToOtlp(ocTracestate *octrace.Span_Tracestate) string {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
+	}
+	return sb.String()
+}
+
+func ocAttrsToOtlp(ocAttrs *octrace.Span_Attributes) (otlpAttrs []*otlpcommon.AttributeKeyValue, droppedCount uint32) {
```

Comment:
We are declaring/allocating the slice below and have explicit return statement. Probably not worth having named returns in this case.

Reply:
The reason is to document the meaning of the second return value (droppedCount). Without a name it is unclear what a uint32 return means.

Reply:
I see. Not a huge deal and it's probably a personal nit for me but I think it might be confusing to mix named returns with explicit returns. May be we can add a comment documenting the return value meaning instead?

Reply:
I'll change to use named returns everywhere.

Reply:
Done.

Reply:
Ah! Don't know how I feel about that now :D

Reply:
Tell me what is confusing.